Python/chapter03/p05_sort_stack/miguelHx.py

```diff
@@ -206,69 +187,46 @@ def test_stack_pop(self):
         self.assertEqual(s._size, 2) # size should now be 2
         self.assertEqual(list(s), [2, 1])
 
-    def test__bool__(self):
-        s = MyStack()
+    def test__bool__(self) -> None:
+        s: MyStack[T] = MyStack()
         self.assertFalse(s)
         s.push(3)
         self.assertTrue(s)
 
 
-def sorted_stack(stack: MyStack[T]) -> MyStack[T]:
+def sorted_stack(stack: MyStack[T]) -> None:
     """This function will take in a stack
-    and return a sorted copy.
-    In order to do this, we need to find the biggest
-    number in an input stack copy, and then push it to the
-    output stack. Repeat until input stack copy is empty.
+    and modify the input stack to be sorted such 
+    that the smallest elements are at the top.
+    Runtime: O(n^2)
+    Space: O(n) where n is the number of elements in the input stack.
 
     Args:
         stack (MyStack): stack of items
-
-    Returns:
-        MyStack: sorted stack of items, with smallest items on top
     """
-    # create empty output stack
-    output_stack: MyStack = MyStack()
-    # make copy of input stack.
-    stack_copy: MyStack = stack.produce_copy()
     # create temporary auxiliary stack
-    aux_stack: MyStack = MyStack()
-    # we will extract max values until stack_copy is empty.
-    while len(stack_copy) > 0:
-        # look for index of max item in stack_copy
-        max_index = 0
-        max_value: T = stack_copy.peek()
-        item: T
-        for i, item in enumerate(stack_copy):
-            if item > max_value:
-                max_value = item
-                max_index = i
-        # biggest values at bottom of output stack
-        output_stack.push(max_value)
-        # next, extract max value and clean up
-        # with the help of aux_stack
-        # We need to remove this value from the stack_copy
-        # In setting up for next max value extraction
-        for i in range(max_index+1):
-            aux_stack.push(stack_copy.pop())
-        # value of interest is now at top of aux_stack
-        # we can discard it and rebuild stack_copy
-        aux_stack.pop()
-        for item in aux_stack:
-            stack_copy.push(aux_stack.pop())
-        # begin next iteration...
-    # done. return sorted stack :D
-    return output_stack
-
-
-class TestSortStack(unittest.TestCase):
-    def test_sort_stack(self):
-        s = MyStack(1, 9, 5, 7, 3, 8)
+    aux_stack: MyStack[T] = MyStack()
+    while stack:
+        t: T = stack.pop()
+        while aux_stack and aux_stack.peek() > t:
+            stack.push(aux_stack.pop())
+        aux_stack.push(t)
```

Comment:
is this really O(n²)? how many operations would be done if this were to receive a stack in reverse order?

Reply:
Yes, in the worst case.
Best case is O(n), which is when we have the input stack already in sorted order with smallest items on top.
Worst case is O(n^2), which is when we have the input stack sorted in reverse order.

Did a small test that prints out number of operations here:  
9be629da04310b85cd3d19dac3e8f6a54e052ef0

When n = 5 in best case, number of operations is 5.
When n = 5 in worst case, number of operations becomes 25.